course6/toyapp-adam/displaylib/scene.cc

```diff
@@ -0,0 +1,67 @@
+#include "scene.h"
+
+#include "surface.h"
+#include "util/keyboard.h"
+#include "util/mouse.h"
+#include <cmath>
+
+namespace Display {
+
+Scene::Scene(Keyboard& keyboard, Mouse& mouse, Surface3D& surface)
+:	mKeyboard{keyboard},
+	mMouse{mouse},
+	mSurface{surface},
+	mTestCube1{{0.0f, 0.0f, 2.5f}, 0.5f, Display::YELLOW},
+	mTestCube2{{1.25f, 0.25f, 2.5f}, 0.5f, Display::GREEN},
+	mTestCube3{{-1.25f, -0.25f, 2.5f}, 0.3f, Display::RED},
+	mTestCube4{{0.0f, 0.75f, 2.5f}, 0.25f, Display::YELLOW},
+	mTestCube5{{0.0f, 1.125f, 2.5f}, 0.125f, Display::YELLOW}
+{}
+
+void Scene::Update(double totalElapsedSeconds)
+{
+	constexpr float camMovementSpeed = 0.01f;
+	constexpr float camRotationSpeed = 0.8f;
+	Vector3D cameraMove;
+
+	if(mKeyboard.IsDown(SDL_SCANCODE_LEFT))
+		cameraMove.x = -camMovementSpeed;
+	if(mKeyboard.IsDown(SDL_SCANCODE_RIGHT))
+		cameraMove.x = +camMovementSpeed;
+	if(mKeyboard.IsDown(SDL_SCANCODE_PAGEUP))
+		cameraMove.y = +camMovementSpeed;
+	if(mKeyboard.IsDown(SDL_SCANCODE_PAGEDOWN))
+		cameraMove.y = -camMovementSpeed;
+	if(mKeyboard.IsDown(SDL_SCANCODE_DOWN))
+		cameraMove.z = -camMovementSpeed;
+	if(mKeyboard.IsDown(SDL_SCANCODE_UP))
+		cameraMove.z = +camMovementSpeed;
+
+	mSurface.camera.transform.Move(cameraMove);
+
+	const Pixel screenDimensions = mSurface.GetDimensions();
```

Comment:
HA! maybe te mouse class should give pointer coordinates in logical units, os that you don't have to do the physical/logical conversion all over your codebase

Reply:
You mean for example a Vector2D in range [-1 to +1]? Sounds good.